Reduce Bridging and Runtime Creation of Strings

LumioseSil opened the pull request:
#### 3969767bf564ae0d40cd2d6ec2f22fac8c83034e
<pre>
Reduce Bridging and Runtime Creation of Strings
<a href="https://bugs.webkit.org/show_bug.cgi?id=251687">https://bugs.webkit.org/show_bug.cgi?id=251687</a>

Reviewed by NOBODY (OOPS!).

In many places, we needlessly bridge for the sake of calling
Objective-C methods and Core Foundation functions where the native
functions/methods work just fine. This PR replaces the bridging with
calling the functions associated with the variable type. This is focused
primarily on strings but some fixes required changing other variables to
CF types or Objective-C NS types.

*Source\WTF\wtf\cocoa\FileSystemCocoa.mm:
*Source\WebCore\bridge\objc\objc_class.mm:
*Source\WebCore\platform\cocoa\MIMETypeRegistryCocoa.mm:
*Source\WebCore\platform\gamepad\cocoa\GameControllerGamepadProvider.mm:
*Source\WebKit\UIProcess\API\Cocoa\WKWebView.mm:
*Source\WebKit\UIProcess\API\Cocoa\_WKSystemPreferences.mm:
*Source\WebKit\UIProcess\API\Cocoa\_WKSystemPreferencesInternal.h:
*Source\WebKit\UIProcess\Cocoa\WebProcessPoolCocoa.mm:
*Source\WebKit\UIProcess\Inspector\mac\WebInspectorUIProxyMac.mm:
*Source\WebKit\UIProcess\WebsiteData\Cocoa\WebsiteDataStoreCocoa.mm:
*Source\WebKit\UIProcess\mac\WKPrintingView.mm:
*Source\WebKitLegacy\mac\Misc\WebNSDataExtras.mm:(_web_capitalizeRFC822HeaderFieldName):
This function was heavily rewritten to call Objective-C functions with
minimal bridging. In addition, unlike the way the function was written
previously, this function frees the backing buffer. However, as a
caveat, and because we cannot change this, said buffer MUST use
malloc/free instead of the standard and established way we allocate and
free temporary buffers.
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3969767bf564ae0d40cd2d6ec2f22fac8c83034e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18615 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42252 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/990 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118651 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/113376 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20082 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/9813 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101769 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115247 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/98187 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/43163 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96933 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/29840 "Found 60 new test failures: accessibility/file-upload-button-stringvalue.html, accessibility/mac/editable-webarea-context-menu-point.html, accessibility/mac/focus-setting-selection-syncronizing-not-clearing.html, animations/restart-after-scroll-nested.html, animations/restart-after-scroll.html, compositing/fixed-with-fixed-layout.html, compositing/fixed-with-main-thread-scrolling.html, compositing/iframes/layout-on-compositing-change.html, compositing/layer-creation/zoomed-clip-intersection.html, compositing/no-compositing-when-full-screen-is-present.html ... (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/84927 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/98401 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11356 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/31183 "Passed tests") | [  ~~🛠 jsc-armv7~~](https://ews-build.webkit.org/#/builders/38/builds/99284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9381 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/12019 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/8115 "Passed tests") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/38/builds/99284 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17385 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/50787 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107267 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13751 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26483 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

aproskuryakov commented on Source/WebKitLegacy/mac/WebView/WebView.mm:
```diff
@@ -8406,26 +8406,32 @@ + (void)_setCacheModel:(WebCacheModel)cacheModel
+#else
+        cfurlCacheDirectory = (__bridge CFStringRef)NSHomeDirectory();
+#endif
+    }
```
There is a whole lot of subtle behavior differences between functions used here. While I don't have a specific problematic scenario to pinpoint, it is more likely than not that we'd break something with this. And given that this is legacy code that is not performance sensitive at all, I'd recommend leaving it as is.

cdumez commented on Source/WebCore/platform/ios/LegacyTileGrid.mm:
```diff
@@ -115,10 +115,11 @@ static bool isFartherAway(const std::pair<double, T>& a, const std::pair<double,
 
-    int bytesToReclaim = int(bytesUsed) - (int(maximumBytes) - bytesNeeded);
-    if (bytesToReclaim <= 0)
+    if (bytesUsed <= maximumBytes - bytesNeeded)
```
This doesn't seem related to the overall PR intention?

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/NSAttributedString.mm:
```diff
@@ -456,7 +456,7 @@ + (void)loadFromHTMLWithData:(NSData *)data options:(NSDictionary<NSAttributedSt
     [self _loadFromHTMLWithOptions:options contentLoader:^WKNavigation *(WKWebView *webView) {
         auto* textEncodingName = dynamic_objc_cast<NSString>(options[NSTextEncodingNameDocumentOption]);
-        auto characterEncoding = static_cast<NSStringEncoding>(dynamic_objc_cast<NSNumber>(options[NSCharacterEncodingDocumentOption]).unsignedIntegerValue);
+        unsigned long characterEncoding = (dynamic_objc_cast<NSNumber>(options[NSCharacterEncodingDocumentOption]).unsignedLongValue);
```
Again, unclear how this relates to the PR initial intention.

cdumez commented on Source/WebKit/UIProcess/Inspector/mac/WebInspectorUIProxyMac.mm:
```diff
@@ -918,15 +918,17 @@ - (IBAction)_popUpButtonAction:(id)sender
+    result.targetProductVersion = static_cast<CFStringRef>(CFDictionaryGetValue(plist, _kCFSystemVersionProductUserVisibleVersionKey));
     result.targetIsSimulator = false;
 
+    CFRelease(plist);
```
We shouldn't do explicit memory management, please use smart pointers.

AZero13 replied:
It is so trivial here though. Like, I am pretty sure a smart pointer here is WORSE because unlike other cases where the memory retention needs change over time, it is pretty clear what this is for and it will only be needed for this length of this function.

Not even worrying about multiple codepaths.

cdumez commented on Source/WebKit/UIProcess/mac/LegacySessionStateCoding.cpp:
```diff
@@ -1133,7 +1133,7 @@ bool decodeLegacySessionState(const uint8_t* bytes, size_t size, SessionState& s
     if (size < sizeof(uint32_t))
         return false;
 
-    uint32_t versionNumber = (bytes[0] << 24) + (bytes[1] << 16) + (bytes[2] << 8) + bytes[3];
```
Again, seems unrelated to this PR initial intention and should be done separately if needed.

cdumez commented on Source/WebKit/UIProcess/mac/WKPrintingView.mm:
```diff
@@ -449,9 +449,9 @@ - (unsigned)_pageForRect:(NSRect)rect
 }
 
-static NSString *linkDestinationName(PDFDocument *document, PDFDestination *destination)
+static CFStringRef createLinkDestinationName(PDFDocument *document, PDFDestination *destination)
```
Should return a `RetainPtr<CFStringRef>`

cdumez commented on Source/WebKit/UIProcess/mac/WKPrintingView.mm:
```diff
@@ -449,9 +449,9 @@ - (unsigned)_pageForRect:(NSRect)rect
+static CFStringRef createLinkDestinationName(PDFDocument *document, PDFDestination *destination)
 {
-    return [NSString stringWithFormat:@"%lu-%f-%f", (unsigned long)[document indexForPage:destination.page], destination.point.x, destination.point.y];
+    return CFStringCreateWithFormat(kCFAllocatorDefault, NULL, CFSTR("%lu-%f-%f"), (unsigned long)[document indexForPage:destination.page], destination.point.x, destination.point.y);
```
Should adoptCF()

cdumez commented on Source/WebKit/UIProcess/mac/WKPrintingView.mm:
```diff
@@ -483,7 +483,9 @@ - (void)_drawPDFDocument:(PDFDocument *)pdfDocument page:(unsigned)page atPoint:
-        CGPDFContextAddDestinationAtPoint(context, (__bridge CFStringRef)linkDestinationName(pdfDocument, destination.get()), destinationPoint);
+        CFStringRef destinationName = createLinkDestinationName(pdfDocument, destination.get());
+        CGPDFContextAddDestinationAtPoint(context, destinationName, destinationPoint);
+        CFRelease(destinationName);
```
No explicit memory management please. We prefer smart pointers.

AZero13 replied:
I understand. However, if it is trivial, then I do not see why not. Though, I do understand. The adoptCF way is better

cdumez commented on Source/WebKit/UIProcess/mac/WKPrintingView.mm:
```diff
@@ -500,7 +502,9 @@ - (void)_drawPDFDocument:(PDFDocument *)pdfDocument page:(unsigned)page atPoint:
-            CGPDFContextSetDestinationForRect(context, (__bridge CFStringRef)linkDestinationName(pdfDocument, destination), transformedRect);
+            CFStringRef destinationName = createLinkDestinationName(pdfDocument, destination);
+            CGPDFContextSetDestinationForRect(context, destinationName, transformedRect);
+            CFRelease(destinationName);
```
ditto.

cdumez commented on Source/WebKitLegacy/mac/Misc/WebNSDataExtras.mm:
```diff
@@ -60,15 +57,16 @@ - (NSString *)_web_capitalizeRFC822HeaderFieldName
                     useUniCharPtr = FALSE;
-                    charPtr = static_cast<char*>(CFAllocatorAllocate(kCFAllocatorDefault, len + 1, 0));
-                    CFStringGetCString(name, charPtr, len+1, kCFStringEncodingISOLatin1);
+                    charPtr = static_cast<char *>(malloc(len + 1));
```
I admit I don't fully understand all the changes to this function and an explanation in the changelog would be helpful (or its own PR). However, we generally use fastMalloc() / fastFree() in WebKit, not malloc()/free().

AZero13 replied:
Actually, this must be an exception; [[NSString alloc] initWithCharactersNoCopy:uniCharPtr length:len freeWhenDone:YES];

freeWhenDone calls free(); There is nothing that can be changed about this.

This has to be done via malloc/free for reasons beyond Webkit's control.

AZero13 replied:
The way it was done before also leaked memory. This portion of the PR is killing two birds with one stone.

cdumez commented on Source/WebKitLegacy/mac/Misc/WebNSDataExtras.mm:
```diff
@@ -136,6 +145,8 @@ - (NSString *)_webkit_guessedMIMETypeForXML
             return nil;
 
         // Skip the "<" and continue.
+        if (p + remaining <= (hit + 1))
```
How does this related to the initial PR intention?

AZero13 replied:
NSUInteger casting to int

cdumez reviewed (changes requested):
Please stick to smaller / easily reviewable changes with detailed commit log explanations.

LumioseSil committed:
Reduce Bridging and Runtime Creation of Strings
https://bugs.webkit.org/show_bug.cgi?id=251687

Reviewed by NOBODY (OOPS!).

In many places, we needlessly bridge for the sake of calling
Objective-C methods and Core Foundation functions where the native
functions/methods work just fine. This PR replaces the bridging with
calling the functions associated with the variable type. This is focused
primarily on strings but some fixes required changing other variables to
CF types or Objective-C NS types.

*Source\WTF\wtf\cocoa\FileSystemCocoa.mm:
*Source\WebCore\bridge\objc\objc_class.mm:
*Source\WebCore\platform\cocoa\MIMETypeRegistryCocoa.mm:
*Source\WebCore\platform\gamepad\cocoa\GameControllerGamepadProvider.mm:
*Source\WebKit\UIProcess\API\Cocoa\WKWebView.mm:
*Source\WebKit\UIProcess\API\Cocoa\_WKSystemPreferences.mm:
*Source\WebKit\UIProcess\API\Cocoa\_WKSystemPreferencesInternal.h:
*Source\WebKit\UIProcess\Cocoa\WebProcessPoolCocoa.mm:
*Source\WebKit\UIProcess\Inspector\mac\WebInspectorUIProxyMac.mm:
*Source\WebKit\UIProcess\WebsiteData\Cocoa\WebsiteDataStoreCocoa.mm:
*Source\WebKit\UIProcess\mac\WKPrintingView.mm:
*Source\WebKitLegacy\mac\Misc\WebNSDataExtras.mm:(_web_capitalizeRFC822HeaderFieldName):
This function was heavily rewritten to call Objective-C functions with
minimal bridging. In addition, unlike the way the function was written
previously, this function frees the backing buffer. However, as a
caveat, and because we cannot change this, said buffer MUST use
malloc/free instead of the standard and established way we allocate and
free temporary buffers.
